projects/wave-core-new/src/lib/users/user.service.ts

```diff
@@ -207,6 +207,26 @@ export class UserService {
 
         return of(session);
     }
+
+    oidcInit() : Observable<AuthCodeRequestURL> {
+        return this.backend.oidcInit();
+    }
+
+    oidcLogin(request: {session_state: string, code: string, state: string}) : Observable<Session> {
+        const result = new Subject<Session>();
+        this.backend
+            .oidcLogin(request)
+            .pipe(mergeMap((response) => this.createSession(response)))
+            .subscribe(
+                (session) => {
+                    this.session$.next(session);
+                    result.next(session);
+                },
+                (error) => result.error(error),
+                () => result.complete(),
+            );
+        return result.asObservable();
+    }
```

Comment:
I guess you need a replaySubject in order to prevent race-conditions.
I've created a helper function here (https://github.com/geo-engine/geoengine-ui/blob/944a795aeb32cc96da765ea35c7bd78e966608f5/projects/wave-core-new/src/lib/project/project.service.ts#L847-L857) because we ran into problems.

But it could be that this is only the case if this emits multiple values :thinking: .

Maybe we can refactor this into a util function and use it also here.

Reply:
I honestly did not think about this, since it is the same for the normal login. FWIW, I don't see the race condition.

Reply:
I guess then it is also wrong for the normal login.

The race is:

1. create the subject
2. set next
3. set completed
4. subscribe to the  subject
5. get completed

…and I guess you would expect to get the session in `next`.
It happens rarely but caused a bug in the project service some months ago.

With the replay subject, the `next` value would be replayed. But you can correct me if I am wrong here.

Reply:
Ok. I guess the issue with the refactoring is session$.next, which is not part of the subscribeAndProvide function.

Does the order of this.session$.next(session) and result.next(session) matter? E.g., the IfLoggedInDirective triggers based on the session$, so I guess there might also be an issue if it is updated and result isn't (or vice versa). In practice, I would expect this to resolve itself - but clearly, I don't yet fully understand when those rxjs events trigger.

Reply:
Added a ReplaySubject for now. 3f6c5b249eddc2a067ffa5c5ffd717cf3b8f0bcc